cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan for upgrading the charm.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan for upgrading the charm.
+        :rtype: Optional[UpgradeStep]
+        """
+        if self.channel != self.next_channel(target):
+            return UpgradeStep(
+                description=(
+                    f"Upgrade '{self.name}' to the new channel: '{self.next_channel(target)}'"
+                ),
+                parallel=parallel,
+                function=async_upgrade_charm,
+                application_name=self.name,
+                channel=self.next_channel(target),
+                model_name=self.model_name,
+            )
+        return None
+
+    def _get_disable_action_managed_plan(self, parallel: bool = False) -> Optional[UpgradeStep]:
+        """Get plan to disable action-managed-upgrade.
+
+        This is used to upgrade as "all-in-one" strategy.
+
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan to disable action-managed-upgrade
+        :rtype: Optional[UpgradeStep]
+        """
+        if self.config.get("action-managed-upgrade", {}).get("value", False):
+            return UpgradeStep(
+                description=(
+                    f"Change charm config of '{self.name}' " "'action-managed-upgrade' to False."
+                ),
+                parallel=parallel,
+                function=async_set_application_config,
+                application_name=self.name,
+                configuration={"action-managed-upgrade": False},
+            )
+        return None
+
+    def _get_workload_upgrade_plan(
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get workload upgrade plan by changing openstack-origin or source.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Workload upgrade plan
+        :rtype: Optional[UpgradeStep]
+        """
+        if self.os_origin != self.new_origin(target):
+            return UpgradeStep(
+                description=(
+                    f"Change charm config of '{self.name}' "
+                    f"'{self.origin_setting}' to '{self.new_origin(target)}'"
+                ),
+                parallel=parallel,
+                function=async_set_application_config,
+                application_name=self.name,
+                configuration={self.origin_setting: self.new_origin(target)},
+            )
+        logger.warning(
+            "App: %s already have %s set to %s",
```

Comment:
WARN here is correct, since if the workload upgrade has already happened we should have hit that HaltUpgradePlanGeneration exception. I'd make the message more prominent:

```suggestion
            "Not triggering the workload upgrade of app %s: %s already set to %s",
```